pkg/kube/krt/collection.go

```diff
@@ -0,0 +1,383 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/kube/controllers"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/maps"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/util/sets"
+)
+
+type manyCollection[I, O any] struct {
+	parent Collection[I]
+	log    *istiolog.Scope
+
+	// mu protects all items grouped below
+	mu              sync.Mutex
+	collectionState multiIndex[I, O]
+	dependencies    sets.String
+	// Stores a map of I -> Things depending on it
+	objectRelations map[Key[I]]dependencies
```

Comment:
objectRelations seems pretty neutral, their is no meaning included about the direction of the reliance on these objects. Maybe `dependents` or `derivedObjects` would inform about the direction of decency.

dependencies == things I depends on
dependents == things that depend on I